### PR TITLE
Add support for NO_PROXY to Mix.Utils.read_path

### PR DESCRIPTION
**Pull request by michaelkschmidt**
This fixes the corner-case where mix needs to access files that are unreachable from the proxy server.

An example is if you have a git repo on the local network, but still need to use the proxy server to fetch the rest of your dependencies.

### CODE REVIEW

**Comment by josevalim** [comment]
@michaelkschmidt thank you!

@ericmj all yours.

File: lib/mix/lib/mix/utils.ex [hunk 1]
```diff
@@ -464,16 +464,30 @@ defmodule Mix.Utils do
   defp proxy_env do
     http_proxy  = System.get_env("HTTP_PROXY")  || System.get_env("http_proxy")
     https_proxy = System.get_env("HTTPS_PROXY") || System.get_env("https_proxy")
+    no_proxy    = System.get_env("NO_PROXY")    || System.get_env("no_proxy")
+                |> no_proxy_list()
```
Comment (ericmj) [inline lib/mix/lib/mix/utils.ex]: Parenthesis are needed around the expression before `|>`. May be better to change to a function call instead and ignore the aligning?

File: lib/mix/lib/mix/utils.ex [hunk 2]
```diff
@@ -464,16 +464,30 @@ defmodule Mix.Utils do
...
+    {proxy_setup(:http, http_proxy, no_proxy), proxy_setup(:https, https_proxy, no_proxy)}
   end
 
-  defp proxy_setup(scheme, proxy) do
+  defp no_proxy_list(nil) do
+    []
+  end
+
+  defp no_proxy_list(no_proxy_str) do
+
```
Comment (ericmj) [inline lib/mix/lib/mix/utils.ex]: ✂️ this line.

**Comment by ericmj** [comment]
@michaelkschmidt Looks good, I just added a few comments.

File: lib/mix/lib/mix/utils.ex [hunk 3]
```diff
@@ -464,16 +464,30 @@ defmodule Mix.Utils do
...
+  defp no_proxy_list(nil) do
+    []
+  end
+
+  defp no_proxy_list(no_proxy_str) do
+    
+    no_proxy_list = String.split(no_proxy_str, ",")
+    for item <- no_proxy_list do
+      to_charlist(item)
+    end
```
Comment (whatyouhide) [inline lib/mix/lib/mix/utils.ex]: Maybe we could refactor all this to be just 

```elixir
Enum.map(String.split(no_proxy_str, ","), &to_charlist/1)
```

or maybe

```elixir
no_proxy_str
|> String.split(",")
|> Enum.map(&to_charlist/1)
```

if we want to be less concise. Wdyt? :)

**Comment by michaelkschmidt** [comment]
@ericmj cool! I wondered about the order of ops for "||" vs "|>", too.  How does this look?

File: lib/mix/lib/mix/utils.ex (hunk 3 above)
Comment (michaelkschmidt) [inline lib/mix/lib/mix/utils.ex]: @whatyouhide thanks for the tips.  I would be cool with the second one if that works for you and @ericmj

Comment (whatyouhide) [inline lib/mix/lib/mix/utils.ex]: @michaelklishin yours works perfectly, sorry I didn't say that explicitly, this was a bit nitpicky but maybe the pipes (`|>`) version could read a bit nicer than the `for` comprehension :)

**Comment by michaelkschmidt** [comment]
@ericmj @whatyouhide any other thoughts?

File: lib/mix/lib/mix/utils.ex [hunk 4]
```diff
@@ -464,16 +464,32 @@ defmodule Mix.Utils do
...
-  defp proxy_setup(scheme, proxy) do
+  defp no_proxy_env() do
+    System.get_env("NO_PROXY") || System.get_env("no_proxy")
+  end
+
+  defp no_proxy_list(nil) do
+    []
+  end
+
+  defp no_proxy_list(no_proxy_str) do
```
Comment (lexmag) [inline lib/mix/lib/mix/utils.ex]: Seems `_str` suffix is superfluous.

**Pull request merged**